File: src/spiders/Emol.ts
import { Paper } from '../classes/Paper'
import { Newspaper } from '../enums'
import { differenceDays } from '../helpers/date'
import { fetchPage } from '../helpers/fetch'

export class Emol {
  private readonly BASE_URL = 'https://newsapi.ecn.cl/NewsApi/emol/buscador/emol?'

  public async run (startDate: Date, endDate: Date): Promise<Paper[]> {
    console.log('Obteniendo noticias desde Emol...')
    console.time('Emol')

    const totalPapers = await this.getTotalPapers()
    const startPaper = await this.getStartPaper(startDate, 0, totalPapers - 1)
    const endPaper = await this.getEndPaper(endDate, 0, startPaper)

    const papersJson = await this.getPaperJson(startPaper, endPaper)
    papersJson.reverse() //* Siempre guardar las noticias mas antiguas primero

    const allPapers: Paper[] = []
    for (const paperJson of papersJson) {
      const paper = new Paper(Newspaper.EMOL, paperJson._source.permalink)
      paper.setEmolData(paperJson._source)
      allPapers.push(paper)
    }

    console.timeEnd('Emol')
    return allPapers
  }

  private async getTotalPapers (): Promise<number> {
    const url = `${this.BASE_URL}size=1&from=0`
    const data: EmolApiResponse = await fetchPage(url)

    return data.hits.total
  }

  private async getStartPaper (startDate: Date, startPaper: number, endPaper: number): Promise<number> {
    if (startPaper >= endPaper - 1) return endPaper

    const midPaper = Math.floor((startPaper + endPaper) / 2)

    const data: EmolApiResponse = await fetchPage(`${this.BASE_URL}size=1&from=${midPaper}`)
    if (data === undefined) return await this.getStartPaper(startDate, startPaper + 1, endPaper)

    const paperDate = new Date(data.hits.hits[0]._source.fechaPublicacion.split('T')[0])

    const diff = differenceDays(startDate, paperDate)
    if (diff === 1) {
      return midPaper
    } else if (diff > 1) {
      return await this.getStartPaper(startDate, startPaper, midPaper)
    } else {
      return await this.getStartPaper(startDate, midPaper, endPaper)
    }
  }

  private async getEndPaper (endDate: Date, startPaper: number, endPaper: number): Promise<number> {
    if (startPaper >= endPaper - 1) return startPaper

    const midPaper = Math.floor((startPaper + endPaper) / 2)

    const data: EmolApiResponse = await fetchPage(`${this.BASE_URL}size=1&from=${midPaper}`)
    if (data === undefined) return await this.getStartPaper(endDate, startPaper, endPaper - 1)

    const paperDate = new Date(data.hits.hits[0]._source.fechaPublicacion.split('T')[0])

    const diff = differenceDays(endDate, paperDate)
    if (diff === -1) {
      return midPaper
    } else if (diff < -1) {
      return await this.getEndPaper(endDate, midPaper, endPaper)
    } else {
      return await this.getEndPaper(endDate, startPaper, midPaper)
    }
  }

  private async getPaperJson (startPaper: number, endPaper: number): Promise<EmolApiHitHit[]> {
    const MAX_SIZE = 500
    const size = startPaper - (endPaper - 1)

    if (size > MAX_SIZE) {
      const hits: EmolApiHitHit[] = []
      for (let i = endPaper; i < startPaper; i += MAX_SIZE) {
        console.log(`URL: ${this.BASE_URL}size=${MAX_SIZE}&from=${i}`)
        const data: EmolApiResponse = await fetchPage(`${this.BASE_URL}size=${MAX_SIZE}&from=${i}`)
        hits.push(...data.hits.hits)
      }
      return hits
    }

    console.log(`URL: ${this.BASE_URL}size=${size}&from=${endPaper}`)
    const data: EmolApiResponse = await fetchPage(`${this.BASE_URL}size=${size}&from=${endPaper}`)
    return data.hits.hits
  }
}

File: src/spiders/emol.py
import logging
import math
from datetime import datetime

from classes.paper import Paper
from services.web_fetcher import WebFetcher
from utils.date_utils import DateUtils


class Emol:
    def __init__(self):
        self.SITE_NAME = "EMOL"
        self.BASE_URL = "https://newsapi.ecn.cl/NewsApi/emol/buscador/emol?"
        self.MAX_PAPERS = 250
        self.fetcher = WebFetcher(delay=10, max_concurrent=1)

        # Configure logging
        logging.basicConfig(level=logging.INFO, format="%(levelname)s: %(message)s")
        self.logger = logging.getLogger(__name__)

    async def run(self, start_date: datetime, end_date: datetime) -> list[Paper]:
        self.logger.info(f"Obteniendo noticias desde {self.SITE_NAME}...")

        total_papers = self.get_total_papers()
        start_paper = self.get_start_paper(start_date, 0, total_papers - 1)
        end_paper = self.get_end_paper(end_date, 0, start_paper)

        urls = self.generate_urls(start_paper, end_paper)
        all_papers = []
        for url in urls:
            papers = self.get_list_papers(url)
            all_papers.extend(filter(lambda p: (p is not None) and (p.date >= start_date and p.date <= end_date), papers))

        return all_papers

    def get_total_papers(self) -> int:
        data = self.fetcher.fetch_json(f"{self.BASE_URL}size=1&from=0")
        if data is None:
            return 0

        return data["hits"]["total"]

    def get_start_paper(self, start_date: datetime, start_paper: int, end_paper: int) -> int:
        if start_paper >= end_paper - 1:
            return end_paper

        mid_paper = math.floor((start_paper + end_paper) / 2)

        data = self.fetcher.fetch_json(f"{self.BASE_URL}size=1&from={mid_paper}")
        if data is None:
            return self.get_start_paper(start_date, start_paper + 1, end_paper)

        paper_date_str = data["hits"]["hits"][0]["_source"]["fechaPublicacion"].split("T")[0]
        paper_date = datetime.strptime(paper_date_str, "%Y-%m-%d")

        diff = DateUtils.diff_days(start_date, paper_date)
        if diff == 1:
            return mid_paper
        elif diff > 1:
            return self.get_start_paper(start_date, start_paper, mid_paper)
        else:
            return self.get_start_paper(start_date, mid_paper, end_paper)

    def get_end_paper(self, end_date: datetime, start_paper: int, end_paper: int) -> int:
        if start_paper >= end_paper - 1:
            return start_paper

        mid_paper = math.floor((start_paper + end_paper) / 2)

        data = self.fetcher.fetch_json(f"{self.BASE_URL}size=1&from={mid_paper}")
        if data is None:
            return self.get_end_paper(end_date, start_paper, end_paper - 1)

        paper_date_str = data["hits"]["hits"][0]["_source"]["fechaPublicacion"].split("T")[0]
        paper_date = datetime.strptime(paper_date_str, "%Y-%m-%d")

        diff = DateUtils.diff_days(end_date, paper_date)
        if diff == -1:
            return mid_paper
        elif diff < -1:
            return self.get_end_paper(end_date, mid_paper, end_paper)
        else:
            return self.get_end_paper(end_date, start_paper, mid_paper)

    def generate_urls(self, start_paper: int, end_paper: int) -> list[str]:
        urls = []
        for i in range(end_paper, start_paper, self.MAX_PAPERS):
            url = f"{self.BASE_URL}size={self.MAX_PAPERS}&from={i}"
            urls.append(url)

        return urls

    def get_list_papers(self, url: str) -> list[Paper]:
        data = self.fetcher.fetch_json(url, is_success=True)
        if data is None:
            return []

        papers = []
        hits = data["hits"]["hits"]
        for hit in hits:
            paper = Paper(self.SITE_NAME, hit["_source"]["permalink"])
            paper.set_emol_data(hit["_source"])
            papers.append(paper)

        return papers
